### PR TITLE
Introduce "exec-runnables-recipe" resolver

### PR DESCRIPTION
**Pull request by clebergnu**
This resolver is somewhat of a hybrid between the "exec-test" and the "runnables-recipe" resolvers.

It runs an executable, and attempts to read from its STDOUT content that will be treated as runnables-recipe JSON content.  If that succeeds, the content will be returned as test resolutions.  This is useful for executable tests or test generators that will output the tests dynamically.

Fixes: https://github.com/avocado-framework/avocado/issues/6009

### CODE REVIEW

File: avocado/plugins/resolvers.py [hunk 1]
```diff
@@ -195,3 +197,84 @@ def resolve(self, reference):
...
+        except (FileNotFoundError, PermissionError) as exc:
+            return ReferenceResolution(
+                reference,
+                ReferenceResolutionResult.NOTFOUND,
+                info=(f'Failure while running running executable "{reference}": {exc}'),
+            )
+
+        content, _ = process.communicate()
+        try:
+            runnables = json.loads(content)
```
Comment (richtja) [inline avocado/plugins/resolvers.py:277]: This whole part from 258 to 280 is duplication of `RunnablesRecipeResolver` code. IMO, it would be better to call the `RunnablesRecipeResolver` or create some Base class for holding the json validation a resolving code.

File: avocado/plugins/resolvers.py [hunk 2]
```diff
@@ -196,3 +198,102 @@ def resolve(self, reference):
...
+            default=False,
+            help_msg=help_msg,
+        )
+
+        help_msg = (
+            "Command line options (space separated) that will be added "
+            "to the executable when executing it as a producer of "
+            "runnables-recipe JSON content."
+        )
+        settings.register_option(
```
Comment (richtja) [inline avocado/plugins/resolvers.py:227]: I am not sure if this is the best solution, what would happen when there would be more than one `run-executables-resolver`, all of them should get the same arguments?

What about have the arguments as part of the reference behind `:`. Like this:
```
$ avocado list --resolver-run-executables examples/nrunner/resolvers/exec_runnables_recipe_kind.sh:tap
```

**Review by richtja — CHANGES_REQUESTED** [review]
Hi @clebergnu, I don't have many comments to the code itself, but I think we should discuss the overall behaviour of exec-runnables-recipe resolver. 

IIUIC, with your solution we can combine `exec-runnables-recipe` and `exec-test` resolvers together, which IMO is not good behaviour, because we could end up in situations where tests are run during resolving phase. I would propose two possible solutions here.

1. If `--resolver-run-executables` option is used all the other resolvers are disabled. This is easy solution, but we will lose the possibility of running multiple test types at once.
2. `--resolver-run-executables` won't be a boolean, but it will be list where user can specify executables which should be directly run by `exec-runnables-recipe` resolver. All the other references will go through normal resolution process. I would prefer this kind of solution.

What do you think, about this problem.

**Comment by clebergnu** [comment]
> Hi @clebergnu, I don't have many comments to the code itself, but I think we should discuss the overall behaviour of exec-runnables-recipe resolver.
> 
> IIUIC, with your solution we can combine `exec-runnables-recipe` and `exec-test` resolvers together, which IMO is not good behaviour, because we could end up in situations where tests are run during resolving phase. I would propose two possible solutions here.
> 

True.  Although I believe the solution we have for that is really the use of different suites.  Users should be aware of the behavior, and control, through the use of different suites, if executables are meant to be run during the resolver phase or not.

> 1. If `--resolver-run-executables` option is used all the other resolvers are disabled. This is easy solution, but we will lose the possibility of running multiple test types at once.

Right.

> 2. `--resolver-run-executables` won't be a boolean, but it will be list where user can specify executables which should be directly run by `exec-runnables-recipe` resolver. All the other references will go through normal resolution process. I would prefer this kind of solution.
> 

I think this is flexible, but adds "special" options and behavior to the overall resolver approach.

> What do you think, about this problem.

Like I said before, I think users should pick the references themselves that should be executed for resolutions, and the ones that should not (that will be executed as tests instead). 

I'll go ahead and better document the behavior and put a note about how users can not (or should not) combine those two different types of executables in a single suite.

File: avocado/plugins/resolvers.py (hunk 2 above)
Comment (clebergnu) [reply avocado/plugins/resolvers.py:227]: > I am not sure if this is the best solution, what would happen when there would be more than one `run-executables-resolver`, all of them should get the same arguments?
> 

Yes, the use case right now requires just a single (the same) argument to all references.

> What about have the arguments as part of the reference behind `:`. Like this:
> 
> ```
> $ avocado list --resolver-run-executables examples/nrunner/resolvers/exec_runnables_recipe_kind.sh:tap
> ```

I believe this stretches a lot the format of references (even though they are free form).  But we could end up with references such as:

```
my-script.sh:--first-argument --second-argument
```

Which would require quoting, escaping, etc.

Also, your proposal reminds me of a very old and long attempt/discussion to be able to support command line arguments in references to `exec-test` (then called `SIMPLE` tests).  It turned out not being a great idea, because of complexity and corner cases that could not be solved (is something an argument? another executable? etc...).

File: avocado/plugins/resolvers.py (hunk 1 above)
Comment (clebergnu) [reply avocado/plugins/resolvers.py:277]: I tried to prevent duplication (the 2nd commit is proof), but I found that trying to not duplicate this code (because of the different resolution informational messages, which I believe are a good thing) turned out to bring more complexity than I thought it was worth saving in LOC.

Also, if you look closely this is not a lot statements, with some statements taking like 6 lines because of formatting.

Let me know if you're OK with the current solution.

File: docs/source/guides/writer/chapters/recipes.rst [hunk 3]
```diff
@@ -76,3 +76,78 @@ That will be parsed by the ``runnables-recipe`` resolver, like in
...
+For security reasons, Avocado won't execute files indiscriminately
+when looking for tests (at the resolution phase). One must set the
+``--resolver-run-executables`` command line option (or the underlying
+``resolver.run_executables`` configuration option) to allow running
+executables at the resolver stage.
+
+.. warning:: It's the user's responsibility to give test references
+             (to be resolved and thus executed) that are well behaved
+             in the sense that they will finish executing quickly,
+             actually execute unintended code (such as running tests),
```
Comment (richtja) [inline docs/source/guides/writer/chapters/recipes.rst:97]: I think you are missing something in here. Didn't you won't so say something like: `in the sense that they will finish executing quickly, and it won't actually execute unintended code`?

File: docs/source/guides/writer/chapters/recipes.rst [hunk 4]
```diff
@@ -76,3 +76,78 @@ That will be parsed by the ``runnables-recipe`` resolver, like in
...
+.. literalinclude:: ../../../../../examples/nrunner/resolvers/exec_runnables_recipe_kind.sh
+
+In order to have those tests resolved as ``tap`` tests, one can run::
+
+  $ avocado list --resolver-run-executables --resolver-exec-arguments tap examples/nrunner/resolvers/exec_runnables_recipe_kind.sh
+
+  tap true-test
+  tap false-test
+
+Behavior of ``exec-runnables-recipe`` and ``exec-test`` resolvers
```
Comment (richtja) [inline docs/source/guides/writer/chapters/recipes.rst:128]: I am not sure if I can agree with this section. Because, with current behaviour, any executable reference will be executed during resolver phase when the `--resolver-run-executables` is used, which sounds self-explanatory, but IMO it might be confusing for users, because exec-test will be executed twice (once by resolver and second as a test). Therefore, we should let users know, that this is not good practice and that they should split those types into separated suite to avoid this behaviour.

File: avocado/plugins/resolvers.py (hunk 1 above)
Comment (richtja) [reply avocado/plugins/resolvers.py:277]: Sure, I am ok with keeping it as it is.

**Review by richtja — CHANGES_REQUESTED** [review]
Hi @clebergnu, thank you for your update. The documentation is definitely better, but I think it still needs to be explicit. It should let users know that their tests might be executed twice with `--resolver-run-executables` and a way how to fix this is to run exec-resolvers and exec-tests in different suites.

File: docs/source/guides/writer/chapters/recipes.rst (hunk 3 above)
Comment (clebergnu) [reply docs/source/guides/writer/chapters/recipes.rst:97]: Yes, true!

File: docs/source/guides/writer/chapters/recipes.rst (hunk 4 above)
Comment (clebergnu) [reply docs/source/guides/writer/chapters/recipes.rst:128]: I've added the following block to the docs:

```
.. warning:: Even though it's possible to have ``exec-test`` and
             ``exec-runnable-recipes`` in the same Avocado test suite
             (for instance in an ``avocado run`` command execution)
             it's not recommended on most cases because ``exec-tests``
             will end up being run at the test resolution phase
             in addition to the test execution phase.  It's
             recommended to use multiple ``avocado run``
             commands or use the Job API and mulitple
             :class:`avocado.core.suite.TestSuite`, one for
             ``exec-runnable-recipes`` with the
             ``resolver.run_executables`` options enabled, and
             another for ``exec-tests`` with that option in its
             default state (disabled).

```

**Review by richtja — APPROVED** [review]
Thanks for the changes, it LGTM

**Comment by richtja** [comment]
@clebergnu, I just noticed that spell-check is failing. Can you please fix that? Thanks.

**Comment by clebergnu** [comment]
> @clebergnu, I just noticed that spell-check is failing. Can you please fix that? Thanks.

Spell checking is happy now! Thanks for the review!
